src/io/bin2text.cpp

```diff
@@ -75,23 +70,59 @@ void dump_domains(const dd::FactorGraph &fg, const std::string &filename) {
 
 void dump_factors(const dd::FactorGraph &fg, const std::string &filename) {
   std::ofstream fout(filename);
+  std::vector<int> vals;
+  std::vector<long> weights;
+  string element;
   for (int i = 0; i < fg.n_factor; ++i) {
     dd::Factor &f = fg.factors[i];
     // FIXME this output is lossy since it drops the f.func_id and
     // f.tmp_variables[*].is_positive
     // variable ids the factor is defined over
-    for (const auto &v : f.tmp_variables) {
-      fout << v.vid;
+    for (long v = 0; v < f.n_variables; ++v) {
+      fout << fg.vifs[f.n_start_i_vif + v].vid;
       fout << field_delim;
     }
     switch (f.func_id) {
       case dd::FUNC_SPARSE_MULTINOMIAL: {
-        // followed by a number of weight ids
-        fout << f.weight_ids.size();
+        // Format: NUM_WEIGHTS [VAR1 VAL ID] [VAR2 VAL ID] ... [WEIGHT ID]
+        // transpose tuples; sort to ensure consistency
+        vals.clear();
+        weights.clear();
+        vals.reserve(f.n_variables * f.weight_ids->size());
+        weights.reserve(f.weight_ids->size());
+        std::map<long, long> ordered(f.weight_ids->begin(), f.weight_ids->end());
+        int w = 0;
+        for (auto it = ordered.begin(); it != ordered.end(); it++) {
+          long key = it->first;
+          long wid = it->second;
+          for (int k = f.n_variables - 1; k >= 0; --k) {
+            int vid = fg.vifs[f.n_start_i_vif + k].vid;
+            dd::Variable &var = fg.variables[vid];
+            int val_idx = key % var.cardinality;
+            int val = var.get_domain_value_at(val_idx);
```

Comment:
Is `get_domain_value_at` and `domain_list` solely for bin2text?  Then let's make these lines in the less critical path do a linear search over the `unordered_map` or construct an inverse map just-in-time.  We put a lot of care not to waste memory with redundant data struct unless it was required by a hot path.  bin2text is and will be used only for human intervention, so we can assume it's gonna stay out of the hot path.